tchap_red_list/__init__.py

```diff
@@ -11,36 +11,211 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-from typing import Any, Dict
+import logging
+from typing import Any, Dict, Optional, Tuple
 
 import attr
-from synapse.module_api import ModuleApi
+from synapse.module_api import (
+    DatabasePool,
+    JsonDict,
+    LoggingTransaction,
+    ModuleApi,
+    UserProfile,
+    cached,
+    run_in_background,
+)
+
+logger = logging.getLogger(__name__)
+
+ACCOUNT_DATA_TYPE = "im.vector.hide_profile"
 
 
 @attr.s(auto_attribs=True, frozen=True)
 class RedListManagerConfig:
-    pass
+    discovery_room: Optional[str] = None
 
 
 class RedListManager:
-    def __init__(self, config: RedListManagerConfig, api: ModuleApi):
+    def __init__(
+        self, config: RedListManagerConfig, api: ModuleApi, setup_db: bool = True
+    ):
         # Keep a reference to the config and Module API
         self._api = api
         self._config = config
 
+        # Register callbacks
+        self._api.register_account_data_callbacks(
+            on_account_data_updated=self.update_red_list_status,
+        )
+
+        self._api.register_spam_checker_callbacks(
+            check_username_for_spam=self.check_user_in_red_list,
+        )
+
+        if setup_db:
+            # Set up the storage layer
+            run_in_background(self._setup_db)
```

Comment:
There's an unlikely race where we try to access the `tchap_red_list` table before `_setup_db` has completed.
Can we add a comment about it for our future selves?

Reply:
Sure can. It makes me sad that we have to run the database setup in the background, but unfortunately I don't see a trivial way to do it another way since it's async code and thus can't be run in `__init__` :/

Reply:
Have opened https://github.com/matrix-org/synapse/issues/12433 to fix this issue since it looks very fixable to me.